modules/ref_arch_setup/tasks/install_pe.sh

```diff
@@ -26,6 +25,76 @@ execute_command() {
     fi
 }
 
+# Sometimes it takes more than one run ...
+
+execute_puppet_until_idempotent() {
+  puppet_command="puppet agent -t"
+
+  # The following could be converted into parameters with defaults ...
+
+  maximum_number_of_puppet_runs=8
+  sleep_seconds_when_puppet_run_in_progress=16
+  retry_when_success_with_failures=false
+
+  for run in $(seq 1 $maximum_number_of_puppet_runs); do
+    echo "Executing: '$puppet_command'"
+
+    # To both capture and output command output ...
+    # exec 5>&1
+    # puppet_command_output=$($puppet_command | tee /dev/fd/5; exit ${PIPESTATUS[0]})
+
+    puppet_command_output=$($puppet_command)
+    puppet_exit_code=$?
+    if echo "$puppet_command_output" | grep -q "Run of Puppet configuration client already in progress"; then
+      puppet_run_in_progress=true
+    else
+      puppet_run_in_progress=false
+    fi
+
+    # 0: The run succeeded with no changes or failures; the system was already in the desired state
+    if [ $puppet_exit_code -eq 0 ]; then
+      break
+    fi
+
+    # 1: The run failed, or wasn't attempted due to another run already in progress.
+    if [ $puppet_exit_code -eq 1 ]; then
+      if [ "$puppet_run_in_progress" = true ]; then
+        echo "Sleeping $sleep_seconds_when_puppet_run_in_progress seconds while waiting for another run already in progress"
+        sleep $sleep_seconds_when_puppet_run_in_progress
+        continue
+      else
+        echo "ERROR: '$puppet_command' failed with exit code: $puppet_exit_code"
+        exit 1
+      fi
+    fi
+
+    # 2: The run succeeded, and some resources were changed.
+    if [ $puppet_exit_code -eq 2 ]; then
+      continue
+    fi
+
+    # 4: The run succeeded, and some resources failed.
+    if [ $puppet_exit_code -eq $success_failures ]; then
```

Comment:
success_failures isn't defined anywhere I can find.

Reply:
Gah, I hand-tested the new function ... but not that exit code.